src/lib/tcp/src/seq.rs

```diff
@@ -0,0 +1,270 @@
+// arbitrary sequence numbers modulo 2^32.
+static_assertions::assert_not_impl_any!(Seq: PartialOrd, Ord, From<u32>, std::ops::Deref);
+
+impl Seq {
+    #[inline]
+    pub fn new(x: u32) -> Self {
+        Self(x)
+    }
+}
+
+impl std::fmt::Debug for Seq {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Display for Seq {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<Seq> for u32 {
+    #[inline]
+    fn from(x: Seq) -> Self {
+        x.0
+    }
+}
+
+impl std::ops::Add<u32> for Seq {
+    type Output = Self;
+
+    fn add(self, offset: u32) -> Self::Output {
+        Self::new(self.0.wrapping_add(offset))
+    }
+}
+
+impl std::ops::Sub<u32> for Seq {
+    type Output = Self;
+
+    fn sub(self, offset: u32) -> Self::Output {
+        Self::new(self.0.wrapping_sub(offset))
+    }
+}
+
+impl std::ops::Sub for Seq {
+    type Output = u32;
+
+    fn sub(self, other: Self) -> Self::Output {
+        self.0.wrapping_sub(other.0)
+    }
+}
+
+impl std::ops::AddAssign<u32> for Seq {
+    fn add_assign(&mut self, offset: u32) {
+        self.0 = self.0.wrapping_add(offset);
+    }
+}
+
+impl std::ops::SubAssign<u32> for Seq {
+    fn sub_assign(&mut self, offset: u32) {
+        self.0 = self.0.wrapping_sub(offset);
+    }
+}
+
+/// A half-open range of sequence numbers modulo 2<sup>32</sup> bounded inclusively by `start` and
+/// exclusively by `end`. The starting position can be greater than the ending position.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub(crate) struct SeqRange {
+    /// Inclusive starting position.
+    pub start: Seq,
+    /// Exclusive ending position.
+    pub end: Seq,
+}
+
+impl std::fmt::Debug for SeqRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.start.fmt(f)?;
+        write!(f, "..")?;
+        self.end.fmt(f)?;
+        Ok(())
+    }
+}
+
+impl SeqRange {
+    #[inline]
+    pub fn new(start: Seq, end: Seq) -> Self {
+        Self { start, end }
+    }
+
+    #[inline]
+    pub fn len(&self) -> u32 {
+        self.end - self.start
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns `true` if the sequence number is contained within this half-open range.
+    #[inline]
+    pub fn contains(&self, seq: Seq) -> bool {
+        seq - self.start < self.end - self.start
+    }
+
+    /// Returns the intersecting range if there is a single intersecting range. Returns `None` if
+    /// there is no intersection, or if there are two intersections. A range will always intersect
+    /// with itself, which also holds for empty ranges. An empty range will intersect with a
+    /// non-empty range if the empty range is contained within the non-empty range.
+    pub fn intersection(&self, other: &Self) -> Option<SeqRange> {
```

Comment:
Optional - since we are using a custom range type here, I wonder if it'd be simpler to disallow empty intervals. Maybe by making it a closed interval instead of half-open.

Reply:
I did think about this initially, but empty ranges end up being commonplace in the tcp code, and I found it simpler just to handle empty and non-empty ranges the same rather than have different code paths to handle the two cases. Using a closed interval also adds other questions like what to do for `[0, u32::MAX].len()`, which would be `u32::MAX + 1` which can't be represented as a u32.